### PR TITLE
Feat/dataset refactor

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This is part 2 of a `chemprop` refactor and deals with reorganizing the underlying data on which MPNNs are trained

## Current status
Aside from being bloated, the current data code has mixed functionality and doesn't adhere to general PyTorch conventions.
- there is no concept of a training datum. This is generally produced by `Dataset.__get__()`, but the current object model relies on data preparation being done by a dataloader
- there is no concept of data collation. Rather, data is collated into a training batch via (1) creating a training dataset, (2), selecting some subset of that dataset: `list[MoleculeDatapoint]`, (3) _reconstructing_ a dataset from that sampled batch, and (4) using that dataset to create a `BatchMolGraph`
- Molecular graph featurization is controlled by the global module state of the `featurization` module. That is, there is no way to simultaneously featurize molecules in different ways, and data dimensionality can quickly and quietly become decoupled from the actual inputs. This can be a nightmare to debug, so the two should be more tightly coupled (to the extent that python allows)
- An additional consequence of this is that there is no way to easily extend molecular graph featurization by writing new featurizers. The only way to do so now is to modify the source code itself, and this is bad.

## Changes
- Data preparation is now logically pipelined in the following manner
  1. create your individual data: `list[MoleculeDatapoint]`
  2. create a featurizer with which to generate molecular graphs of those data: `MoleculeFeaturizer()`
  3. create a dataset: `MolGraphDataset(data, featurizer)`
- A `MoleculeFeaturizer` is further composed of both an `AtomFeaturizer` and a `BondFeaturizer`. These can easily be extended by a client for custom functionality.
- Data getting is now done (I would argue) more naturally. When getting an index from a dataset, a tuple is returned containing all inputs* an MPNN needs at train time
- Data loading is also more natural, the tuples from above are collated into a single batch, forming a `BatchMolGraph` for input features and all other features/descriptors/targets are stacked so that the model only ever sees a completely usable batch
- a `MolGraph` is now just a data structure (`NamedTuple`, specifically) rather than the more complex object it was previously. This allows the separation of "data" from "logic". I.e., we no longer couple the featurization logic and the storage of the data together. The benefits here are:
  - simplicity: it is _crystal clear_ now what precisely a `MolGraph` is: just a container for some data in a structured format
  - scalability: when preparing data across multiple workers, the logic is only shipped to the worker once rather than once *for each* data point
  - inspectability: using a `MolGraph` in client code it is now possible to see what attributes are there more clearly
- data is now separated into two types: `Reaction` and `Molecule`. These two types of data are prepared and utilized differently, and the previous code treated them the same through some obtuse abstraction. The abstraction still remains (i.e., they're both amenable as input an MPNN in the same way,) but their data preparation has been logically separated so that a client knowingly decides when to treat their input data as molecule- or reaction-type data

## Questions
There are some bugs in this PR that will be addressed in a later one that is to be opened soon. I'm only splitting up into separate PRs so that folks can follow the process along.

## Checklist
- [x] linted with flake8?
- [ ] (if appropriate) unit tests added? **some** but these will be added in a future PR

### CODE REVIEW

File: chemprop/data/v2/datasets.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            for d in self.data:
+                d.atom_features = scaler.transform(d.raw_atom_features)
+        elif scale_bond_features:
+            for d in self.data:
+                d.bond_features = scaler.transform(d.raw_bond_features)
+        else:
+            for d in self.data:
+                d.features = scaler.transform(d.raw_features.reshape(1, -1))[0]
+
+        return scaler
```
Comment (shihchengli) [inline chemprop/data/v2/datasets.py:229]: I know the extra atom/bond features are currently not supported for reactions. This capability is expected to be supported in the future. Hence, I would suggest moving this part to MolGraphDataset.

**Review by shihchengli — COMMENTED** [review]
Good work on refactoring the way to featurize molecules and load data. I roughly went through the changes you made and made some minor review comments. The code structure is definitely more clear and more efficient. As for the `featurization`, will you open a door to let users easily extend molecular graph featurization in some way?

File: chemprop/data/v2/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+class MolGraphDataLoader(DataLoader):
+    """A `MolGraphDataLoader` is a PyTorch `DataLoader` for loading a `MolGraphDataset`
+
+    Parameters
+    ----------
+    dataset : MoleculeDataset
+        The `MoleculeDataset` containing the molecules to load.
+    batch_size : int, default=50
+        the batch size to load
+    num_workers : int, default=0
```
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/data/v2/dataloader.py:69]: no- that just means loading data in the main process. Were you thinking of something specific?

File: chemprop/models/v2/encoders/reaction.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    @classmethod
+    def from_mol_encoder_args(
+        cls, n_mols: int, shared: bool = False, *args, **kwargs
+    ) -> ReactionEncoder:
+        if not shared:
+            encoders = [molecule_encoder(*args, **kwargs) for _ in range(n_mols)]
+        else:
+            encoders = [molecule_encoder(*args, **kwargs)]
+
+        return cls(encoders, n_mols, shared)
```
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/models/v2/encoders/reaction.py:91]: it's removed in a later PR

File: chemprop/data/v2/datasets.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/data/v2/datasets.py:229]: We can move it over when the features are actually supported. Until then, I don't want to add capability for something that's not actually usable

**Comment by davidegraff** [comment]
The new code has added that extensibility for free. There are two main avenues with which a client could customize `MolGraph` featurization:
1. define a custom `{Molecule,Reaction}Featurizer`. The data preparation code only requires an object that satisfies the underlying call signature `Union[Mol, tuple[Mol, Mol]] -> Optional[Array] -> Optional[Array] -> MolGraph`
2. define custom `{Atom,Bond}Featurizer`s. These can then be supplied to the already existing `{Molecule,Reaction}Featurizer` classes. As long as the custom classes satisfy the parent `{Atom,Bond}Featurizer` interface, they will work just fine within the existing `{Molecule,Reaction}Featurizer` code.

File: chemprop/data/v2/dataloader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/data/v2/dataloader.py:69]: I didn't know what will happen if setting it as 0. Thank you for the clarification, that makes sense to me now.

**Pull request merged**